src/Algebra/Module/Construct/Nagata.agda

```diff
@@ -0,0 +1,170 @@
+
+  *ᴺ-identity : Identity 1ᴺ _*ᴺ_
+  *ᴺ-identity = lᴺ , rᴺ
+    where
+      open ≈-Reasoning setoidᴹ
+      lᴺ : LeftIdentity 1ᴺ _*ᴺ_
+      lᴺ (r , m) = (*ᴿ-identityˡ r) , (begin
+        (1ᴿ *ₗ m +ᴹ 0ᴹ *ᵣ r) ≈⟨ +ᴹ-cong (*ₗ-identityˡ m) (*ᵣ-zeroˡ r) ⟩
+        (m +ᴹ 0ᴹ) ≈⟨ +ᴹ-identityʳ m ⟩
+        m ∎)
+      rᴺ : RightIdentity 1ᴺ _*ᴺ_
+      rᴺ (r , m) = (*ᴿ-identityʳ r) , (begin
+        r *ₗ 0ᴹ +ᴹ m *ᵣ 1ᴿ ≈⟨ +ᴹ-cong (*ₗ-zeroʳ r) (*ᵣ-identityʳ m) ⟩
+        0ᴹ +ᴹ m ≈⟨ +ᴹ-identityˡ m ⟩
+        m ∎)
+
+  *ᴺ-cong : Congruent₂ _*ᴺ_
+  *ᴺ-cong (r₁ , m₁) (r₂ , m₂) = *ᴿ-cong r₁ r₂ , +ᴹ-cong (*ₗ-cong r₁ m₂) (*ᵣ-cong m₁ r₂)
+
+  *ᴺ-assoc : Associative _*ᴺ_
+  *ᴺ-assoc (r₁ , m₁) (r₂ , m₂) (r₃ , m₃) = *ᴿ-assoc r₁ r₂ r₃ , (begin
+    (r₁ *ᴿ r₂) *ₗ m₃ +ᴹ (r₁ *ₗ m₂ +ᴹ m₁ *ᵣ r₂) *ᵣ r₃
+      ≈⟨ +ᴹ-cong (*ₗ-assoc r₁ r₂ m₃) (*ᵣ-distribʳ r₃ (r₁ *ₗ m₂) (m₁ *ᵣ r₂)) ⟩
+    r₁ *ₗ (r₂ *ₗ m₃) +ᴹ ((r₁ *ₗ m₂) *ᵣ r₃ +ᴹ (m₁ *ᵣ r₂) *ᵣ r₃)
+       ≈⟨ +ᴹ-congˡ (+ᴹ-congʳ (*ₗ-*ᵣ-assoc r₁ m₂ r₃)) ⟩
+    r₁ *ₗ (r₂ *ₗ m₃) +ᴹ (r₁ *ₗ (m₂ *ᵣ r₃) +ᴹ (m₁ *ᵣ r₂) *ᵣ r₃)
+      ≈⟨ symᴹ (+ᴹ-assoc (r₁ *ₗ (r₂ *ₗ m₃)) (r₁ *ₗ (m₂ *ᵣ r₃)) ((m₁ *ᵣ r₂) *ᵣ r₃)) ⟩
+    (r₁ *ₗ (r₂ *ₗ m₃) +ᴹ r₁ *ₗ (m₂ *ᵣ r₃)) +ᴹ (m₁ *ᵣ r₂) *ᵣ r₃
+      ≈⟨ +ᴹ-cong (symᴹ (*ₗ-distribˡ r₁ (r₂ *ₗ m₃) (m₂ *ᵣ r₃))) (*ᵣ-assoc m₁ r₂ r₃) ⟩
+    r₁ *ₗ (r₂ *ₗ m₃ +ᴹ m₂ *ᵣ r₃) +ᴹ m₁ *ᵣ (r₂ *ᴿ r₃) ∎)
+    where
+      open ≈-Reasoning setoidᴹ
+
+  *ᴺ-distrib-+ᴺ : _*ᴺ_ DistributesOver _+ᴺ_
+  *ᴺ-distrib-+ᴺ = lᴺ , rᴺ
+    where
+      open ≈-Reasoning setoidᴹ
+      open Consequences setoidᴹ
+      +ᴹ-middleFour = comm∧assoc⇒middleFour +ᴹ-cong +ᴹ-comm +ᴹ-assoc
+      lᴺ : _*ᴺ_ DistributesOverˡ _+ᴺ_
+      lᴺ (r₁ , m₁) (r₂ , m₂) (r₃ , m₃) = *ᴿ-distribˡ-+ᴿ r₁ r₂ r₃ , (begin
+        r₁ *ₗ (m₂ +ᴹ m₃) +ᴹ m₁ *ᵣ (r₂ +ᴿ r₃)
+          ≈⟨ +ᴹ-cong (*ₗ-distribˡ r₁ m₂ m₃) (*ᵣ-distribˡ m₁ r₂ r₃) ⟩
+        (r₁ *ₗ m₂ +ᴹ r₁ *ₗ m₃) +ᴹ (m₁ *ᵣ r₂ +ᴹ m₁ *ᵣ r₃)
+          ≈⟨ +ᴹ-middleFour (r₁ *ₗ m₂) (r₁ *ₗ m₃) (m₁ *ᵣ r₂) (m₁ *ᵣ r₃) ⟩
+        (r₁ *ₗ m₂ +ᴹ m₁ *ᵣ r₂) +ᴹ (r₁ *ₗ m₃ +ᴹ m₁ *ᵣ r₃) ∎)
+      rᴺ : _*ᴺ_ DistributesOverʳ _+ᴺ_
+      rᴺ (r₁ , m₁) (r₂ , m₂) (r₃ , m₃) = *ᴿ-distribʳ-+ᴿ r₁ r₂ r₃ , (begin
+        (r₂ +ᴿ r₃) *ₗ m₁ +ᴹ (m₂ +ᴹ m₃) *ᵣ r₁
+          ≈⟨ +ᴹ-cong (*ₗ-distribʳ m₁ r₂ r₃) (*ᵣ-distribʳ r₁ m₂ m₃) ⟩
+        (r₂ *ₗ m₁ +ᴹ r₃ *ₗ m₁) +ᴹ (m₂ *ᵣ r₁ +ᴹ m₃ *ᵣ r₁)
+          ≈⟨ +ᴹ-middleFour (r₂ *ₗ m₁) (r₃ *ₗ m₁) (m₂ *ᵣ r₁) (m₃ *ᵣ r₁) ⟩
+        (r₂ *ₗ m₁ +ᴹ m₂ *ᵣ r₁) +ᴹ (r₃ *ₗ m₁ +ᴹ m₃ *ᵣ r₁) ∎)
+
+
+------------------------------------------------------------------------
+-- The Fundamental Lemma, due to Nagata
+
+  isRingᴺ : IsRing _≈ᴺ_ _+ᴺ_ _*ᴺ_ -ᴺ_ 0ᴺ  1ᴺ
+  isRingᴺ = record
+    { +-isAbelianGroup = +ᴺ-isAbelianGroup
+    ; *-cong = *ᴺ-cong
+    ; *-assoc = *ᴺ-assoc
+    ; *-identity = *ᴺ-identity
+    ; distrib = *ᴺ-distrib-+ᴺ
+    }
+
+------------------------------------------------------------------------
+-- Bundle
+
+infixl 4 _⋉_
+
+_⋉_ : (R : Ring r ℓr) (M : Bimodule R R m ℓm) → Ring (r ⊔ m) (ℓr ⊔ ℓm)
```

Comment:
So with the top-level module being parameterised, you can still have this definition. For starters I would rename it to `ringᴺ : Ring (r ⊔ m) (ℓr ⊔ ℓm)` as the main definition so that it can be used when the top-level module is opened applied. Then you can define a simple alias `_⋉_ = ringᴺ` so that it can be used when the the top-level module is opened unapplied.

Reply:
Ah, that's nice. I confess I have tended to stumble over module-level scope and infixes before, so end up working as I have done here. But I'll try this!

Reply:
Now done, but not sure if I did it correctly the way you suggested. And what about the potential for "unsolved metas" in using the `_⋉_` notation for the module being opened unapplied?

Reply:
Why would using an infix name as opposed to a standard name result in unsolved metas that weren't there before?

Reply:
Fair point!